The Setter interface only works on struct pointer fields, not the struct itself

```go
package test

import (
	"github.com/creasty/defaults"
	"testing"
)

type Main struct {
	MainInt int `default:"-"`
	*Other  `default:"{}"`
}

type Other struct {
	OtherInt int `default:"-"`
}

func (s *Main) SetDefaults() {
	if defaults.CanUpdate(s.MainInt) {
		s.MainInt = 1
	}
}

func (s *Other) SetDefaults() {
	if defaults.CanUpdate(s.OtherInt) {
		s.OtherInt = 1
	}
}

func TestDefaultsSetter(t *testing.T) {
	main := &Main{}
	defaults.Set(main)
	if main.OtherInt != 1 {
		t.Errorf("expected 1 for OtherInt, got %d", main.OtherInt)
	}
	if main.MainInt != 1 {
		t.Errorf("expected 1 for MainInt, got %d", main.MainInt)
	}
}
```
Which outputs:
```
=== RUN   TestDefaultsSetter
    TestDefaultsSetter: dev_test.go:36: expected 1 for MainInt, got 0
--- FAIL: TestDefaultsSetter (0.00s)
FAIL

Process finished with exit code 1

```

--- a/defaults.go
+++ b/defaults.go
@@ -133,13 +133,11 @@ func setField(field reflect.Value, defaultVal string) error {
 			}
 			field.Set(ref.Elem().Convert(field.Type()))
 		case reflect.Struct:
-			ref := reflect.New(field.Type())
 			if defaultVal != "" && defaultVal != "{}" {
-				if err := json.Unmarshal([]byte(defaultVal), ref.Interface()); err != nil {
+				if err := json.Unmarshal([]byte(defaultVal), field.Addr().Interface()); err != nil {
 					return err
 				}
 			}
-			field.Set(ref.Elem())
 		case reflect.Ptr:
 			field.Set(reflect.New(field.Type().Elem()))
 		}
@@ -150,13 +148,10 @@ func setField(field reflect.Value, defaultVal string) error {
 		setField(field.Elem(), defaultVal)
 		callSetter(field.Interface())
 	case reflect.Struct:
-		ref := reflect.New(field.Type())
-		ref.Elem().Set(field)
-		if err := Set(ref.Interface()); err != nil {
+		if err := Set(field.Addr().Interface()); err != nil {
 			return err
 		}
-		callSetter(ref.Interface())
-		field.Set(ref.Elem())
+		callSetter(field.Addr().Interface())
 	case reflect.Slice:
 		for j := 0; j < field.Len(); j++ {
 			if err := setField(field.Index(j), defaultVal); err != nil {
@@ -176,6 +171,10 @@ func shouldInitializeField(field reflect.Value, tag string) bool {
 	switch field.Kind() {
 	case reflect.Struct:
 		return true
+	case reflect.Ptr:
+		if field.Interface() != nil && field.Elem().Kind() == reflect.Struct {
+			return true
+		}
 	case reflect.Slice:
 		return field.Len() > 0 || tag != ""
 	}

--- a/defaults.go
+++ b/defaults.go
@@ -151,7 +151,6 @@ func setField(field reflect.Value, defaultVal string) error {
 		if err := Set(field.Addr().Interface()); err != nil {
 			return err
 		}
-		callSetter(field.Addr().Interface())
 	case reflect.Slice:
 		for j := 0; j < field.Len(); j++ {
 			if err := setField(field.Index(j), defaultVal); err != nil {

--- a/defaults.go
+++ b/defaults.go
@@ -38,7 +38,7 @@ func Set(ptr interface{}) error {
 			}
 		}
 	}
-
+	callSetter(ptr)
 	return nil
 }
 

--- a/defaults.go
+++ b/defaults.go
@@ -172,14 +172,14 @@ func shouldInitializeField(field reflect.Value, tag string) bool {
 	case reflect.Struct:
 		return true
 	case reflect.Ptr:
-		if field.Interface() != nil && field.Elem().Kind() == reflect.Struct {
+		if !field.IsNil() && field.Elem().Kind() == reflect.Struct {
 			return true
 		}
 	case reflect.Slice:
 		return field.Len() > 0 || tag != ""
 	}
 
-	return (tag != "")
+	return tag != ""
 }
 
 // CanUpdate returns true when the given value is an initial value of its type

--- a/defaults_test.go
+++ b/defaults_test.go
@@ -389,3 +389,20 @@ func TestCanUpdate(t *testing.T) {
 		}
 	}
 }
+
+type Child struct {
+	Name string `default:"Tom"`
+	Age  int    `default:"20"`
+}
+
+type Parent struct {
+	Child *Child
+}
+
+func TestPointerStructMember(t *testing.T) {
+	m := Parent{Child: &Child{Name: "Jim"}}
+	Set(&m)
+	if m.Child.Age != 20 {
+		t.Errorf("20 is expected")
+	}
+}

--- a/defaults_test.go
+++ b/defaults_test.go
@@ -406,3 +406,35 @@ func TestPointerStructMember(t *testing.T) {
 		t.Errorf("20 is expected")
 	}
 }
+
+type Main struct {
+	MainInt int `default:"-"`
+	*Other  `default:"{}"`
+}
+
+type Other struct {
+	OtherInt int `default:"-"`
+}
+
+func (s *Main) SetDefaults() {
+	if CanUpdate(s.MainInt) {
+		s.MainInt = 1
+	}
+}
+
+func (s *Other) SetDefaults() {
+	if CanUpdate(s.OtherInt) {
+		s.OtherInt = 1
+	}
+}
+
+func TestDefaultsSetter(t *testing.T) {
+	main := &Main{}
+	Set(main)
+	if main.OtherInt != 1 {
+		t.Errorf("expected 1 for OtherInt, got %d", main.OtherInt)
+	}
+	if main.MainInt != 1 {
+		t.Errorf("expected 1 for MainInt, got %d", main.MainInt)
+	}
+}

--- a/go.mod
+++ b/go.mod
@@ -1,0 +1,3 @@
+module github.com/creasty/defaults
+
+go 1.11

